### ISSUE TITLE
Blocks: rescue or always not run when error comes from a run_once task

### ISSUE THREAD

**Issue by Gugli** [ansible/ansible#16937]
##### ISSUE TYPE
- Bug Report
##### ANSIBLE VERSION

```
ansible 2.1.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

Out of the box config on Ubuntu14
##### OS / ENVIRONMENT

Ansible runs on Ubuntu14, targeted machines on Ubuntu16.
##### SUMMARY

Using `run_once` on a task inside a block prevents `rescue` and `always` sections to be correctly triggered.
##### STEPS TO REPRODUCE

Run this playbook on any inventory file. This is basically the sample from [the official documentation](http://docs.ansible.com/ansible/playbooks_blocks.html#block-error-handling-example) with just a `run_once` added

```
# Contents of bug_blocks.yml
- hosts: all
  gather_facts: false
  tasks:
  - block:
      - debug: msg='I execute normally'
      - command: /bin/false
        run_once: true
      - debug: msg='I never execute, cause ERROR!'
    rescue:
      - debug: msg='I caught an error'
      - command: /bin/false
      - debug: msg='I also never execute :-('
    always:
      - debug: msg="this always executes"
```
##### EXPECTED RESULTS

I was expecting 

```
$ ansible-playbook bug_blocks.yml --inventory-file inventories/

PLAY [all] *********************************************************************

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "i execute normally"
}

TASK [command] *****************************************************************
fatal: [localhost]: FAILED! => {"changed": true, "cmd": ["/bin/false"], "delta": "0:00:00.002056", "end": "2016-08-03 17:04:07.156094", "failed": true, "rc": 1, "start": "2016-08-03 17:04:07.154038", "stderr": "", "stdout": "", "stdout_lines": [], "warnings": []}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "I caught an error"
}

TASK [command] *****************************************************************
fatal: [localhost]: FAILED! => {"changed": true, "cmd": ["/bin/false"], "delta": "0:00:00.001853", "end": "2016-08-03 17:04:07.988953", "failed": true, "rc": 1, "start": "2016-08-03 17:04:07.987100", "stderr": "", "stdout": "", "stdout_lines": [], "warnings": []}

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "this always executes"
}

NO MORE HOSTS LEFT *************************************************************

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=1
TASK: command ----------------------------------------------------------- 0.79s
TASK: command ----------------------------------------------------------- 0.40s
TASK: debug ------------------------------------------------------------- 0.11s
TASK: debug ------------------------------------------------------------- 0.09s
TASK: debug ------------------------------------------------------------- 0.07s

Playbook finished: Wed Aug  3 17:04:08 2016, 5 total tasks.  0:00:01 elapsed.
```
##### ACTUAL RESULTS

```
$ ansible-playbook bug_blocks.yml --inventory-file inventories/

PLAY [all] *********************************************************************

TASK [debug] *******************************************************************
ok: [localhost] => {
    "msg": "i execute normally"
}

TASK [command] *****************************************************************
fatal: [localhost]: FAILED! => {"changed": true, "cmd": ["/bin/false"], "delta": "0:00:00.001284", "end": "2016-08-03 17:03:52.951110", "failed": true, "rc": 1, "start": "2016-08-03 17:03:52.949826", "stderr": "", "stdout": "", "stdout_lines": [], "warnings": []}

NO MORE HOSTS LEFT *************************************************************

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0

TASK: command ----------------------------------------------------------- 0.71s
TASK: debug ------------------------------------------------------------- 0.09s

Playbook finished: Wed Aug  3 17:26:37 2016, 2 total tasks.  0:00:00 elapsed.
```
##### Related issues

This might be related to #16254, but the steps to reproduce are so different I though it best to make a new Issue report.

**Comment by Gugli** [comment]
Might also be somewhat related to #16050.

**Comment by MadVikingGod** [comment]
I can confirm this bug.  It does not happen on v2.1.0.0-1, but everything later does not work correctly.

I have narrowed it down to fbec2d9692cdbbddda1f6ecb68bbca67e3b783b4  linear.py lines 364 and 365.

Reverting this part of that patch will fix this but, but I think will also undo the fix for #16937.

A real fix for both of these issues is to not treat run_once as any_error_fatal, and have a different logic path if it fails. This would mark all of the hosts in that play as failed, and then run the appropriate recovery logic.

**Comment by jctanner** [comment]
cc @jimi-c

**Comment by jimi-c** [comment]
@Gugli / @MadVikingGod the above branch is a first attempt at fixing this. Still needs some work but I think the basics are there. I still need to:
- go through our other strategy plugins and base strategy to make sure at least in our code we're using integer results values now
- my logic for the handler stuff (and vs. or) is probably not right, but haven't tested it yet (shippable is running on it)
- only the first host in the list executes the rescue - no clue as to why yet because the rest are marked as failed and the PlayIterator should have them all at the same point too.

**Comment by jimi-c** [comment]
Hrm, and it looks like I've got some other bit of logic wrong.. this test succeeded but there was a non-zero exit code:

https://app.shippable.com/runs/57a4dfa5d640fc0d00f90f97/5/console

**Comment by jimi-c** [comment]
<pre>
$ ansible -m ping localhost
localhost | SUCCESS => {
    "changed": false, 
    "ping": "pong"
}
$ echo $?
1
</pre>

**Comment by pporada-gl** [comment]
I hit this today on `ansible 2.1.1.0` without the run_once.

**Comment by jimi-c** [comment]
The branch above is working and passing tests, if anyone else would like to give it a test before I merge it into devel.

**Comment by jimi-c** [comment]
# Closing This Ticket

Hi!

We believe the above commit should resolve this problem for you. This will also be included in the next release.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!

**Comment by jimi-c** [comment]
# Closing This Ticket

Hi!

We believe the above commit should resolve this problem for you. This will also be included in the next release.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
